clouddrift/signal.py

```diff
@@ -234,6 +234,190 @@ def cartesian_to_rotary(
     return wp, wn
 
 
+def ellipse_parameters(
+    xa: Union[np.ndarray, xr.DataArray],
+    ya: Union[np.ndarray, xr.DataArray],
+) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+    """Return the instantaneous parameters of a modulated elliptical signal from its analytic Cartesian signals.
+
+    Parameters
+    ----------
+    xa : array_like
+        Complex-valued analytic signal for first Cartesian component (zonal, east-west).
+    ya : array_like
+        Complex-valued analytic signal for second Cartesian component (meridional, north-south).
+
+    Returns
+    -------
+    kappa : np.ndarray
+        Ellipse root-mean-square amplitude.
+    lambda : np.ndarray
+        Ellipse linearity between -1 and 1, or departure from circular motion (lambda=0).
+    theta : np.ndarray
+        Ellipse orientation in radian.
+    phi : np.ndarray
+        Ellipse phase in radian.
+
+    Examples
+    --------
+
+    To obtain the ellipse parameters from a pair of real-valued signals (x, y):
+
+    >>> kappa, lambda, theta, phi = ellipse_parameters(analytic_signal(x), analytic_signal(y))
+
+    Raises
+    ------
+    ValueError
+        If the input arrays do not have the same shape.
+
+    References
+    ----------
+    Lilly JM, Olhede SC. 2010 Bivariate instantaneous frequency and bandwidth.
+    IEEE T. Signal Proces. 58, 591–603. (10.1109/TSP.2009.2031729).
+
+    See Also
+    --------
+    :func:`modulated_ellipse_signal`, :func:`analytic_signal`, :func:`rotary_to_cartesian`, :func:`cartesian_to_rotary`
+
+    """
+
+    # u and v arrays must have the same shape.
+    if not xa.shape == ya.shape:
+        raise ValueError("xa and ya must have the same shape.")
+
+    # initialize empty array for z; for future extension to 3D
+    # calculation below should hold once za is input
+    za = np.zeros_like(xa)
+
+    # calculate normal vector to ellipse plane
+    nx = np.imag(ya) * np.real(za) - np.imag(za) * np.real(ya)
+    ny = -(np.imag(xa) * np.real(za) - np.imag(za) * np.real(xa))
+    nz = np.imag(xa) * np.real(ya) - np.imag(ya) * np.real(xa)
+
+    denom = np.sqrt(nx**2 + ny**2 + nz**2)
+    nx = nx / denom
+    ny = ny / denom
+    nz = nz / denom
+
+    alpha = np.imag(np.log(np.sqrt(-1) * nx - ny))
+    beta = np.imag(np.log(nz + np.sqrt(-1) * np.sqrt(nx**2 + ny**2)))
+
+    x = np.cos(-alpha) * xa - np.sin(-alpha) * ya
+    y = np.sin(-alpha) * xa + np.cos(-alpha) * ya
+    z = za
+    x = np.cos(-beta) * x - np.sin(-beta) * y
+    y = np.sin(-beta) * x + np.cos(-beta) * y
+    z = z
+
+    X = np.abs(x)
+    Y = np.abs(y)
+    phix = np.angle(x)
+    phiy = np.angle(y)
+
+    phia = (phix + phiy + np.pi / 2) / 2
+    phid = (phix - phiy - np.pi / 2) / 2
+
+    P = 0.5 * np.sqrt(X**2 + Y**2 + 2 * X * Y * np.cos(2 * phid))
+    N = 0.5 * np.sqrt(X**2 + Y**2 - 2 * X * Y * np.cos(2 * phid))
```

Comment:
Minor nitpick, I'd pick one style between `x / 2` and `0.5 * x` and be consistent, for readability.